### PR TITLE
Feature/contact tracing output

### PR DESCRIPTION
**Pull request by emely3h**
## Related Issue
- So far the contact tracing worked like this: The contact tracer can type in a email adress and if there are possible contacts to this email adress the contact tracer can download all possible contacts in one excel sheet. The contact tracer had to separate manually 3 different types of contacts (students, lecturers and employees of the university, guests). 

## Proposed Changes
- Instead of downloading all contacts in one excel sheet the contact tracer can now download three different excel sheets
- One for all students, one for all lecturers and employees of the university and one for all guests
- Also some additional columns are added (adress, ...)

## Additional Info
- The code in the ContactTracingController class from line 110-130 is suboptimal.
- Currently it is not checked if guests type in their adress in the correct format (straße hausnummer plz ort). To avoid an exception if the address is not in the correct format a empty string is returned
- Would be better to check in advance in the check in page if the adress is in the correct format but thats an other issue :)

## Checklist

- [x] Is the code formatted correctly?
- [x] Unnecessary comments removed?
- [x] Printout statements removed?
- [x] If you made backend changes: Did all tests pass? Do you need to adjust some tests or write new ones?
- [x] If you made frontend changes: Did you test the UI on different devices/ browsers (Firefox, Chrome, Safari, different smartphone sizes)
- [x] Is your code easy to understand or do you need to insert some comments?
- [x] Explain the issue for the reviewer and your steps to solve the issue => makes it a lot more easy for the reviewer
- [x] Label your pull request (frontend/ backend/ testing/ styling/ sql/...)
- [ ] Is the open source message included at the very beginning of every source code file?
- [x] Are all imports organized, i.e. are the imports up to date and have all package imports been removed?

### CODE REVIEW

**Comment by oliverhummel** [comment]
Unfortunately, I get an ArrayOutOfBoundsException, which seems to occur only when an external guest gets into the contact tracing. Please check and ideally add a test case!

2021-09-20 23:10:56.923 ERROR 68178 --- [0.0-8080-exec-8] d.h.i.ct.controller.ErrorController      : java.lang.ArrayIndexOutOfBoundsException: 1
java.lang.ArrayIndexOutOfBoundsException: 1
	at de.hs_mannheim.informatik.ct.controller.ContactTracingController.lambda$30(ContactTracingController.java:102)
	at de.hs_mannheim.informatik.ct.controller.ContactTracingController.lambda$45(ContactTracingController.java:181)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at de.hs_mannheim.informatik.ct.controller.ContactTracingController.doSearch(ContactTracingController.java:184)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:197)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:141)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:894)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1060)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:962)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:652)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:327)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:115)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:126)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:149)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:218)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:212)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:103)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:89)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:132)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:110)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:80)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:55)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:336)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:211)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:183)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:346)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:887)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1684)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)

File: src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java [hunk 1]
```diff
@@ -64,14 +64,73 @@
...
+            new TracingColumn("Straße", contact -> ""),
+            new TracingColumn("Hausnummer", contact -> ""),
+            new TracingColumn("Plz", contact -> ""),
+            new TracingColumn("Ort", contact -> ""),
+            new TracingColumn("Tel.", contact -> "")
+    );
+    private final List<TracingColumn> tracingColumnsGuests = Arrays.asList(
+            new TracingColumn("EMail-Adresse", contact -> contact.getContact().getEmail()),
+            new TracingColumn("Nachname", contact -> contact.getContact().getName().split(" ")[0]),
+            new TracingColumn("Vorname", contact -> contact.getContact().getName().split(" ")[1]),
```
Comment (oliverhummel) [inline src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java:102]: I think, the problem lies in the assumption that there is always a first an a last name entered, which is not enforced in any way. Hence, I would leave this in one column together.

Moreover, I just changed the order to Vorname Nachname in the input field today as I find it strange to do it the other way round.

File: src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java [hunk 2]
```diff
@@ -140,4 +237,26 @@ public String doSearch(@RequestParam String email, Model model) {
...
+     * @param contacts List of all possible contacts of the target email
+     * @param type     String of the filter type (students/ staff/ guests)
+     * @return filtered List of Contacts
+     */
+    private List<Contact<?>> filterContactList(Collection<Contact<?>> contacts, String type) {
+        List<Contact<?>> filteredList = new ArrayList<>();
+        for (Contact contact : contacts) {
+            if (type.equals("students") && contact.getContact().getEmail().contains("@stud.hs-mannheim.de")) {
+                filteredList.add(contact);
+            } else if (type.equals("staff") && contact.getContact().getEmail().contains("@hs-mannheim.de")) {
```
Comment (oliverhummel) [inline src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java:254]: Please make sure to add addresses with @lba.hs-mannheim (Lehrbeauftragte) here as well. Cf. PR #245

**Comment by emely3h** [comment]
Thanks, I put the columns for name and adress back together. I also noticed the issue with the adress input field, I wrote an issue about it last week but haven't got around to it yet, but I'll get to it today.

File: src/main/resources/templates/tracing/contactList.html [hunk 3]
```diff
@@ -9,31 +9,86 @@
...
-                </tr>
-                </thead>
-
-                <tr th:each="row: ${tableValues}">
-                    <td th:each="cell: ${row}" th:text="${cell}"/>
-                </tr>
-            </table>
+                Zur Weitergabe an das Gesundheitsamt können die Daten in Form von Excel Dateien heruntergeladen werden.
+                Dabei kann jeweils eine seperate Excel-Datei für Studenten, Lehrende/Hochschulangestellte und Gäste
+                heruntergeladen werden oder alle Kontaktpersonen gesammelt in einer Exceldatei.
```
Comment (oliverhummel) [inline src/main/resources/templates/tracing/contactList.html:23]: I can't see a link for all contacts in one file? Remove this part of the sentence?

File: src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java [hunk 4]
```diff
@@ -64,14 +64,43 @@
...
+    private final List<TracingColumn> tracingColumnsStudents = Arrays.asList(
+            new TracingColumn("EMail-Adresse", contact -> contact.getContact().getEmail()),
+            new TracingColumn("Mtknr.", contact -> contact.getContact().getEmail().split("@")[0]),
+            new TracingColumn("Raum/Veranstaltung", Contact::getContactLocation),
+            new TracingColumn("Datum", contact -> dateFormatter.format(contact.getTargetVisit().getStartDate())),
+            new TracingColumn("Anmeldung Ziel", contact -> timeFormatter.format(contact.getTargetVisit().getStartDate())),
+            new TracingColumn("Anmeldung Kontakt", contact -> timeFormatter.format(contact.getContactVisit().getStartDate())),
+            new TracingColumn("Abmeldung Ziel", contact -> timeFormatter.format(contact.getTargetVisit().getEndDate())),
+            new TracingColumn("Abmeldung Kontakt", contact -> timeFormatter.format(contact.getContactVisit().getEndDate())),
+            new TracingColumn("Geburtsdatum", contact -> ""),
```
Comment (oliverhummel) [inline src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java:76]: Why is there a date of birth in the table? It is never entered anywhere nor is it a legal requirement, is it? Hence, I would remove it?

Comment (oliverhummel) [reply src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java:76]: What I miss in the columns is the phone number though. And I think this is required by law, isn't it?

File: src/main/resources/templates/tracing/contactList.html (hunk 3 above)
Comment (emely3h) [reply src/main/resources/templates/tracing/contactList.html:23]: Yes, removed it

File: src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java (hunk 4 above)
Comment (emely3h) [reply src/main/java/de/hs_mannheim/informatik/ct/controller/ContactTracingController.java:76]: Not shure, mary told me which columns to create, she wanted the date of birth in every sheet and the phone number just for guests. I will adress it in the meeting today.

**Pull request merged**